File: WalletKitJava/WalletKitBRD/src/main/java/com/blockset/walletkit/brd/Network.java
/*
 * Created by Michael Carrara <michael.carrara@breadwallet.com> on 7/1/19.
 * Copyright (c) 2019 Breadwinner AG.  All right reserved.
*
 * See the LICENSE file at the project root for license information.
 * See the CONTRIBUTORS file at the project root for a list of contributors.
 */
package com.blockset.walletkit.brd;

import androidx.annotation.Nullable;

import com.blockset.walletkit.nativex.cleaner.ReferenceCleaner;
import com.blockset.walletkit.nativex.WKAddressScheme;
import com.blockset.walletkit.nativex.WKCurrency;
import com.blockset.walletkit.nativex.WKNetwork;
import com.blockset.walletkit.nativex.WKNetworkFee;
import com.blockset.walletkit.nativex.WKSyncMode;
import com.blockset.walletkit.AddressScheme;
import com.blockset.walletkit.NetworkType;
import com.blockset.walletkit.WalletManagerMode;
import com.google.common.base.Optional;
import com.google.common.base.Supplier;
import com.google.common.base.Suppliers;
import com.google.common.primitives.UnsignedInteger;
import com.google.common.primitives.UnsignedLong;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.Set;

import static com.google.common.base.Preconditions.checkState;

/* package */
final class Network implements com.blockset.walletkit.Network {

    static Network create(WKNetwork core, boolean needTake) {
        Network network = new Network(
                (needTake ? core.take() : core)
        );
        ReferenceCleaner.register(network, core::give);
        return network;
    }

    /* package */
    static Network from(com.blockset.walletkit.Network network) {
        if (network == null) {
            return null;
        }

        if (network instanceof Network) {
            return (Network) network;
        }

        throw new IllegalArgumentException("Unsupported network instance");
    }

    private final WKNetwork core;

    private final Supplier<String> uidsSupplier;
    private final Supplier<String> nameSupplier;
    private final Supplier<Boolean> isMainnetSupplier;
    private final Supplier<NetworkType> typeSupplier;
    private final Supplier<Currency> currencySupplier;
    private final Supplier<Set<Currency>> currenciesSupplier;

    private Network(WKNetwork core) {
        this.core = core;

        uidsSupplier = Suppliers.memoize(core::getUids);
        nameSupplier = Suppliers.memoize(core::getName);
        isMainnetSupplier = Suppliers.memoize(core::isMainnet);
        typeSupplier = Suppliers.memoize(() -> Utilities.networkTypeFromCrypto(core.getCanonicalType()));
        currencySupplier = Suppliers.memoize(() -> Currency.create(core.getCurrency()));

        currenciesSupplier = Suppliers.memoize(() -> {
            Set<Currency> currencies = new HashSet<>();
            UnsignedLong count = core.getCurrencyCount();
            for (UnsignedLong i = UnsignedLong.ZERO; i.compareTo(count) < 0; i = i.plus(UnsignedLong.ONE)) {
                currencies.add(Currency.create(core.getCurrency(i)));
            }
            return currencies;
        });
    }

    @Override
    public String getUids() {
        return uidsSupplier.get();
    }

    @Override
    public String getName() {
        return nameSupplier.get();
    }

    @Override
    public boolean isMainnet() {
        return isMainnetSupplier.get();
    }

    @Override
    public NetworkType getType() {
        return typeSupplier.get();
    }

    @Override
    public UnsignedLong getHeight() {
        return core.getHeight();
    }

    @Override
    public UnsignedInteger getConfirmationsUntilFinal() {
        return core.getConfirmationsUntilFinal();
    }

    @Override
    public Optional<NetworkPeer> createPeer(String address, UnsignedInteger port, @Nullable String publicKey) {
        return NetworkPeer.create(this, address, port, publicKey);
    }

    @Override
    public Currency getCurrency() {
        return currencySupplier.get();
    }

    @Override
    public Set<Currency> getCurrencies() {
        return new HashSet<>(currenciesSupplier.get());
    }

    @Override
    public Optional<Currency> getCurrencyByCode(String code) {
        for (Currency currency: getCurrencies()) {
            if (code.equals(currency.getCode())) {
                return Optional.of(currency);
            }
        }
        return Optional.absent();
    }

    @Override
    public Optional<Currency> getCurrencyByIssuer(String issuer) {
        String issuerLowercased = issuer.toLowerCase(Locale.ROOT);
        for (Currency currency: getCurrencies()) {
            if (issuerLowercased.equals(currency.getIssuer().orNull())) {
                return Optional.of(currency);
            }
        }
        return Optional.absent();
    }

    @Override
    public List<? extends NetworkFee> getFees() {
        List<NetworkFee> fees = new ArrayList<>();
        for (WKNetworkFee fee: core.getFees()) {
            fees.add(NetworkFee.create(fee));
        }
        return fees;
    }

    @Override
    public NetworkFee getMinimumFee() {
        NetworkFee minimumFee = null;
        for (NetworkFee fee: getFees()) {
            if (minimumFee == null || fee.getConfirmationTimeInMilliseconds().compareTo(minimumFee.getConfirmationTimeInMilliseconds()) > 0) {
                minimumFee = fee;
            }
        }
        return minimumFee;
    }

    @Override
    public boolean hasCurrency(com.blockset.walletkit.Currency currency) {
        return core.hasCurrency(Currency.from(currency).getCoreBRCryptoCurrency());
    }

    @Override
    public Optional<Unit> baseUnitFor(com.blockset.walletkit.Currency currency) {
        if (!hasCurrency(currency)) {
            return Optional.absent();
        }
        return core.getUnitAsBase(Currency.from(currency).getCoreBRCryptoCurrency()).transform(Unit::create);
    }

    @Override
    public Optional<Unit> defaultUnitFor(com.blockset.walletkit.Currency currency) {
        if (!hasCurrency(currency)) {
            return Optional.absent();
        }
        return core.getUnitAsDefault(Currency.from(currency).getCoreBRCryptoCurrency()).transform(Unit::create);
    }

    @Override
    public Optional<Set<? extends com.blockset.walletkit.Unit>> unitsFor(com.blockset.walletkit.Currency currency) {
        if (!hasCurrency(currency)) {
            return Optional.absent();
        }

        Set<Unit> units = new HashSet<>();

        WKCurrency currencyCore = Currency.from(currency).getCoreBRCryptoCurrency();
        UnsignedLong count = core.getUnitCount(currencyCore);

        for (UnsignedLong i = UnsignedLong.ZERO; i.compareTo(count) < 0; i = i.plus(UnsignedLong.ONE)) {
            Optional<Unit> unit = core.getUnitAt(currencyCore, i).transform(Unit::create);
            if (!unit.isPresent()) {
                return Optional.absent();
            }

            units.add(unit.get());
        }

        return Optional.of(units);
    }

    @Override
    public Optional<Boolean> hasUnitFor(com.blockset.walletkit.Currency currency, com.blockset.walletkit.Unit unit) {
        return unitsFor(currency).transform(input -> input.contains(unit));
    }

    public void addCurrency(com.blockset.walletkit.Currency currency,
                            com.blockset.walletkit.Unit baseUnit,
                            com.blockset.walletkit.Unit defaultUnit) {
        checkState (baseUnit.hasCurrency(currency));
        checkState (defaultUnit.hasCurrency(currency));
        if (!this.hasCurrency(currency)) {
            getCoreBRCryptoNetwork().addCurrency (
                    Currency.from(currency).getCoreBRCryptoCurrency(),
                    Unit.from(baseUnit).getCoreBRCryptoUnit(),
                    Unit.from(defaultUnit).getCoreBRCryptoUnit()
            );
        }
    }

    public void addUnitFor(com.blockset.walletkit.Currency currency,
                           com.blockset.walletkit.Unit unit) {
        checkState (unit.hasCurrency(currency));
        checkState (this.hasCurrency(currency));
        if (!this.hasUnitFor (currency, unit).isPresent()) {
            getCoreBRCryptoNetwork().addCurrencyUnit(
                    Currency.from(currency).getCoreBRCryptoCurrency(),
                    Unit.from(unit).getCoreBRCryptoUnit()
            );
        }
    }

    @Override
    public Optional<Address> addressFor(String address) {
        return Address.create(address, this);
    }

    @Override
    public AddressScheme getDefaultAddressScheme() {
        return Utilities.addressSchemeFromCrypto (getCoreBRCryptoNetwork().getDefaultAddressScheme());
    }

    @Override
    public List<AddressScheme> getSupportedAddressSchemes() {
        List<AddressScheme> schemes = new ArrayList<>();
        for (WKAddressScheme c: getCoreBRCryptoNetwork().getSupportedAddressSchemes())
            schemes.add(Utilities.addressSchemeFromCrypto(c));
        return schemes;
    }

    @Override
    public boolean supportsAddressScheme(AddressScheme addressScheme) {
        return getCoreBRCryptoNetwork().supportsAddressScheme(Utilities.addressSchemeToCrypto(addressScheme));
    }

    @Override
    public WalletManagerMode getDefaultWalletManagerMode() {
        return Utilities.walletManagerModeFromCrypto (getCoreBRCryptoNetwork().getDefaultSyncMode());
    }

    @Override
    public List<WalletManagerMode> getSupportedWalletManagerModes() {
        List<WalletManagerMode> modes = new ArrayList<>();
        for (WKSyncMode m: getCoreBRCryptoNetwork().getSupportedSyncModes())
            modes.add (Utilities.walletManagerModeFromCrypto(m));
        return modes;
    }

    @Override
    public boolean supportsWalletManagerMode(WalletManagerMode mode) {
        return getCoreBRCryptoNetwork().supportsSyncMode (Utilities.walletManagerModeToCrypto(mode));
    }

    @Override
    public boolean requiresMigration() {
        return getCoreBRCryptoNetwork().requiresMigration();
    }

    @Override
    public String toString() {
        return getName();
    }

    @Override
    public boolean equals(Object object) {
        if (this == object) {
            return true;
        }

        if (!(object instanceof Network)) {
            return false;
        }

        Network network = (Network) object;
        return getUids().equals(network.getUids());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getUids());
    }

    /* package */
    void setHeight(UnsignedLong height) {
        core.setHeight(height);
    }

    void setVerifiedBlockHashAsString(String hash) {
        core.setVerifiedBlockHashAsString(hash);
    }

    /* package */
    void setFees(List<NetworkFee> fees) {
        checkState(!fees.isEmpty());
        List<WKNetworkFee> cryptoFees = new ArrayList<>(fees.size());
        for (NetworkFee fee: fees) {
            cryptoFees.add(fee.getCoreBRCryptoNetworkFee());
        }
        core.setFees(cryptoFees);
    }

    /* package */
    WKNetwork getCoreBRCryptoNetwork() {
        return core;
    }

    public static Optional<Network> findBuiltin (String name) {
        return WKNetwork.findBuiltin(name)
                .transform(Network::new);
    }
}

File: WalletKitKotlin/src/jvmMain/kotlin/Network.kt
package com.breadwallet.core

import com.breadwallet.corenative.cleaner.ReferenceCleaner
import com.breadwallet.corenative.crypto.BRCryptoAddressScheme
import com.breadwallet.corenative.crypto.BRCryptoNetwork
import com.breadwallet.corenative.crypto.BRCryptoPeer
import com.breadwallet.corenative.crypto.BRCryptoSyncMode
import com.google.common.primitives.UnsignedInteger
import com.google.common.primitives.UnsignedLong
import kotlinx.io.core.Closeable

actual class Network internal constructor(
        internal val core: BRCryptoNetwork
) : Closeable {

    init {
        ReferenceCleaner.register(core, ::close)
    }

    internal actual val uids: String = core.uids

    actual val name: String = core.name

    actual val type: NetworkType
        get() = NetworkType.fromCoreInt(core.canonicalType.toCore())

    actual val isMainnet: Boolean = core.isMainnet

    actual var height: ULong
        get() = core.height.toLong().toULong()
        internal set(value) {
            core.height = UnsignedLong.valueOf(value.toLong())
        }

    actual var fees: List<NetworkFee>
        get() = core.fees.map(::NetworkFee)
        set(value) {
            require(value.isNotEmpty())
            core.fees = value.map(NetworkFee::core)
        }

    actual val minimumFee: NetworkFee
        get() = checkNotNull(fees.maxBy(NetworkFee::timeIntervalInMilliseconds))

    actual val confirmationsUntilFinal: UInt
        get() = core.confirmationsUntilFinal.toByte().toUInt()

    actual fun createPeer(address: String, port: UShort, publicKey: String?): NetworkPeer? =
            BRCryptoPeer.create(
                    core,
                    address,
                    UnsignedInteger.valueOf(port.toLong()),
                    publicKey
            ).orNull()?.run(::NetworkPeer)

    actual val currency: Currency by lazy { Currency(core.currency) }

    actual val currencies: Set<Currency> by lazy {
        (0 until core.currencyCount.toLong())
                .map { core.getCurrency(UnsignedLong.valueOf(it)) }
                .map(::Currency)
                .toSet()
    }

    actual val defaultAddressScheme: AddressScheme
        get() = AddressScheme.fromCoreInt(core.defaultAddressScheme.toCore().toUInt())

    actual val supportedAddressSchemes: List<AddressScheme>
        get() = core.supportedAddressSchemes.map {
            AddressScheme.fromCoreInt(it.toCore().toUInt())
        }

    actual val supportedWalletManagerModes: List<WalletManagerMode>
        get() = core.supportedSyncModes.map { WalletManagerMode.fromCoreInt(it.toCore().toUInt()) }

    actual val defaultWalletManagerMode: WalletManagerMode
        get() = WalletManagerMode.fromCoreInt(core.defaultSyncMode.toCore().toUInt())

    actual fun supportsWalletManagerMode(mode: WalletManagerMode): Boolean {
        return core.supportsSyncMode(BRCryptoSyncMode.fromCore(mode.core.toInt()))
    }

    actual fun supportsAddressScheme(addressScheme: AddressScheme): Boolean {
        return core.supportsAddressScheme(BRCryptoAddressScheme.fromCore(addressScheme.core.toInt()))
    }

    actual fun currencyByCode(code: String): Currency? =
            currencies.firstOrNull { it.code == code }

    actual fun currencyByIssuer(issuer: String): Currency? {
        val issuerLowerCase = issuer.toLowerCase()
        return currencies.firstOrNull { currency ->
            currency.issuer?.toLowerCase() == issuerLowerCase
        }
    }

    actual fun hasCurrency(currency: Currency): Boolean =
            core.hasCurrency(currency.core)

    actual fun baseUnitFor(currency: Currency): CUnit? {
        if (!hasCurrency(currency)) return null
        val cryptoUnit = core.getUnitAsBase(currency.core).orNull() ?: return null
        return CUnit(cryptoUnit)
    }

    actual fun defaultUnitFor(currency: Currency): CUnit? {
        if (!hasCurrency(currency)) return null
        val cryptoUnit = core.getUnitAsDefault(currency.core).orNull() ?: return null
        return CUnit(cryptoUnit)
    }

    actual fun unitsFor(currency: Currency): Set<CUnit>? {
        if (!hasCurrency(currency)) return null
        return (0 until core.getUnitCount(currency.core).toLong())
                .map { checkNotNull(core.getUnitAt(currency.core, UnsignedLong.valueOf(it)).orNull()) }
                .map { CUnit(it) }
                .toSet()
    }

    actual fun hasUnitFor(currency: Currency, unit: CUnit): Boolean? =
            unitsFor(currency)?.contains(unit)

    actual fun addressFor(string: String): Address? {
        return Address.create(string, this)
    }

    actual fun addCurrency(currency: Currency, baseUnit: CUnit, defaultUnit: CUnit) {
        require(baseUnit.hasCurrency(currency))
        require(defaultUnit.hasCurrency(currency))
        if (!hasCurrency(currency)) {
            core.addCurrency(currency.core, baseUnit.core, defaultUnit.core)
        }
    }

    actual fun addUnitFor(currency: Currency, unit: CUnit) {
        require(unit.hasCurrency(currency))
        require(hasCurrency(currency))
        if (hasUnitFor(currency, unit) == null) {
            core.addCurrencyUnit(currency.core, unit.core)
        }
    }

    actual fun requiresMigration(): Boolean =
            core.requiresMigration()

    actual override fun hashCode(): Int = uids.hashCode()
    actual override fun equals(other: Any?): Boolean =
            other is Network && core.uids == other.uids

    actual override fun toString(): String = name

    actual override fun close() {
        core.give()
    }

    actual companion object {
        internal actual fun installBuiltins(): List<Network> =
                BRCryptoNetwork.installBuiltins().map(::Network)

        actual fun findBuiltin(uids: String): Network? =
                BRCryptoNetwork.findBuiltin(uids).orNull()?.run(::Network)
    }
}
